src/poseidon/mod.rs

```diff
@@ -0,0 +1,360 @@
+                res.push(F::rand(rng));
+            }
+            ark.push(res);
+        }
+
+        ark
+    }
+}
+
+impl<F: PrimeField> CryptographicSponge for PoseidonSponge<F> {
+    type Parameters = PoseidonParameters<F>;
+
+    fn new(params: &Self::Parameters) -> Self {
+        // Requires F to be Alt_Bn128Fr
+        let full_rounds = params.full_rounds;
+        let partial_rounds = params.partial_rounds;
+        let alpha = params.alpha;
+
+        let mds = params.mds.clone();
+
+        let ark = params.ark.to_vec();
+
+        let rate = 2;
+        let capacity = 1;
+        let state = vec![F::zero(); rate + capacity];
+        let mode = PoseidonSpongeMode::Absorbing {
+            next_absorb_index: 0,
+        };
+
+        Self {
+            full_rounds,
+            partial_rounds,
+            alpha,
+            ark,
+            mds,
+
+            state,
+            rate,
+            capacity,
+            mode,
+        }
+    }
+
+    fn absorb(&mut self, input: &impl Absorb) {
+        let elems = input.to_sponge_field_elements_as_vec::<F>();
+        if elems.is_empty() {
+            return;
+        }
+
+        match self.mode {
+            PoseidonSpongeMode::Absorbing { next_absorb_index } => {
+                let mut absorb_index = next_absorb_index;
+                if absorb_index == self.rate {
+                    self.permute();
+                    absorb_index = 0;
+                }
+                self.absorb_internal(absorb_index, elems.as_slice());
+            }
+            PoseidonSpongeMode::Squeezing {
+                next_squeeze_index: _,
+            } => {
+                self.permute();
+                self.absorb_internal(0, elems.as_slice());
+            }
+        };
+    }
+
+    fn squeeze_bytes(&mut self, num_bytes: usize) -> Vec<u8> {
+        let usable_bytes = (F::Params::CAPACITY / 8) as usize;
+
+        let num_elements = (num_bytes + usable_bytes - 1) / usable_bytes;
+        let src_elements = self.squeeze_native_field_elements(num_elements);
+
+        let mut bytes: Vec<u8> = Vec::with_capacity(usable_bytes * num_elements);
+        for elem in &src_elements {
+            let elem_bytes = elem.into_repr().to_bytes_le();
+            bytes.extend_from_slice(&elem_bytes[..usable_bytes]);
+        }
+
+        bytes.truncate(num_bytes);
+        bytes
+    }
+
+    fn squeeze_bits(&mut self, num_bits: usize) -> Vec<bool> {
+        let usable_bits = F::Params::CAPACITY as usize;
+
+        let num_elements = (num_bits + usable_bits - 1) / usable_bits;
+        let src_elements = self.squeeze_native_field_elements(num_elements);
+
+        let mut bits: Vec<bool> = Vec::with_capacity(usable_bits * num_elements);
+        for elem in &src_elements {
+            let elem_bits = elem.into_repr().to_bits_le();
+            bits.extend_from_slice(&elem_bits[..usable_bits]);
+        }
+
+        bits.truncate(num_bits);
+        bits
+    }
+}
+
```

Comment:
Are we able to/should we override the default implementation of `squeeze_field_elements` from the trait to call `squeeze_native_field_elements` when possible?

Reply:
I also want to do this. Main problem is that when overriding `squeeze_field_elements<F: PrimeField>` (for example, in `Poseidon`), rust compiler believes `F` and the field used by sponge is different. Only way I can come up with is to convert the squeezed native field elements to bytes and convert back. Shall we do that?

Reply:
Even that conversion seems more direct and a bit more efficient than the default approach for `squeeze_field_elements`. Also, ensuring that `squeeze_field_elements` for the native field returns the same thing as `squeeze_native_field_elements` would prevent weird inconsistencies/bugs that might occur later on. Changing the function to directly call the `squeeze_native_field_elements` would be the most direct way of ensuring consistent behavior.

Reply:
Will do it today

Reply:
I've added a commit implementing this! Please check